app/backends/workflow-orchestration.ts

```diff
@@ -526,138 +537,162 @@ export async function handleWorkItemUpdate(
-          workItem.scrollID = scrollID;
-          await maybeQueueQueryCmrWorkItem(tx, workItem, logger);
-        } else {
-          // Failed to create the next work items - fail the job rather than leaving it orphaned
-          // in the running state
-          logger.error('The work item update should have contained results to queue a next work item, but it did not.');
-          const message = 'Harmony internal failure: could not create the next work items for the request.';
-          await completeJob(tx, job, JobStatus.FAILED, logger, message);
-        }
-      } else {
-        // Finished with the chain for this granule
-        if (status != WorkItemStatus.FAILED) {
-          await addJobLinksForFinishedWorkItem(tx, job, results, logger);
+      const continueProcessing = await handleFailedWorkItems(tx, job, workItem, thisStep, status, logger, errorMessage);
+      if (continueProcessing) {
+        const nextWorkflowStep = await getWorkflowStepByJobIdStepIndex(
+          tx, workItem.jobID, workItem.workflowStepIndex + 1,
+        );
+        if (nextWorkflowStep && (status !== WorkItemStatus.FAILED || nextWorkflowStep?.isBatched)) {
+          createdWorkItem = await createNextWorkItems(
+            tx,
+            nextWorkflowStep,
+            logger,
+            workItem,
+            status,
+            allWorkItemsForStepComplete,
+            results,
+            outputItemSizes,
+          );
         }
-        // If all granules are finished mark the job as finished
-        job.completeBatch(thisStep.workItemCount);
-        if (allWorkItemsForStepComplete) {
-          const finalStatus = await getFinalStatusForJob(tx, job);
-          await completeJob(tx, job, finalStatus, logger);
-        } else {
-          // Special case to pause the job as soon as any single granule completes when in the previewing state
-          if (job.status === JobStatus.PREVIEWING) {
-            job.pause();
+
+        if (nextWorkflowStep && status === WorkItemStatus.SUCCESSFUL) {
+          if (results && results.length > 0) {
+            // set the scrollID for the next work item to the one we received from the update
+            workItem.scrollID = scrollID;
+            await maybeQueueQueryCmrWorkItem(tx, workItem, logger);
+          } else {
+            // Failed to create the next work items when there should be work items.
+            // Fail the job rather than leaving it orphaned in the running state
+            logger.error('The work item update should have contained results to queue a next work item, but it did not.');
+            const message = 'Harmony internal failure: could not create the next work items for the request.';
+            await completeJob(tx, job, JobStatus.FAILED, logger, message);
+          }
+        } else if (!nextWorkflowStep || allWorkItemsForStepComplete) {
+          // Finished with the chain for this granule
```

Comment:
I don't think this comment is true if `nextWorkflowStep` is `true`, which it might be.

Reply:
A successful work item would have entered the prior if block when `nextWorkflowStep` is `true`, so the only way to be in this block having  `nextWorkflowStep=true` would be if the item failed, in which case we are done processing that granule, so the comment is correct.

Reply:
OK, 'finished with the chain' implies success to me, but maybe not